cpp/src/io/json/read_json.cu

```diff
@@ -41,6 +42,57 @@ namespace cudf::io::json::detail {
 
 namespace {
 
+class compressed_host_buffer_source final : public datasource {
+ public:
+  explicit compressed_host_buffer_source(std::unique_ptr<datasource> src, compression_type comptype)
+    : _comptype{comptype}
+  {
+    auto dbuf  = src->host_read(0, src->size());
+    _ch_buffer = std::vector<std::uint8_t>(dbuf->data(), dbuf->data() + dbuf->size());
+    if (comptype == compression_type::GZIP || comptype == compression_type::ZIP ||
+        comptype == compression_type::SNAPPY) {
+      _decompressed_ch_buffer_size = estimate_uncompressed_size(_comptype, _ch_buffer);
+      _decompressed_buffer.resize(0);
+    } else {
+      _decompressed_buffer         = decompress(_comptype, _ch_buffer);
+      _decompressed_ch_buffer_size = _decompressed_buffer.size();
+    }
+  }
+
+  size_t host_read(size_t offset, size_t size, uint8_t* dst) override
+  {
+    auto decompressed_hbuf = decompress(_comptype, _ch_buffer);
```

Comment:
I think we always want to save the result in `_decompressed_buffer`. We cannot guarantee any read pattern and saving the decompressed buffer on first read avoid repeated calls to `decompress`.

Reply:
Good point, fixed this `host_read` to use the same memoization logic as `host_read(offset, size)`